### DISCUSSION TITLE
node numbers need shift in LinearNodalConstraint and Output by -1? Abaqus mesh related?

### DISCUSSION THREAD

**Discussion by arndflatten** [idaholab/moose#26638] [Q&A General] [answered]
Dear Moose-Forum,

I like to use Abaqus-meshes via the 'FileMeshGenerator'. However, we noticed when testing the 'LinearNodalConstraint' for a simple problem, that using the node labels for parameters primary and secondary_node_ids cannot be used as input in the mesh file.

We understand that abaqus meshes should have both node and element labels consecutively numbered starting from label 1 (actually starting with node label 0 is not valid any more, which makes sense).
Also, the flag  allow_renumbering=false was set.

Still the node numbers that one needs to specify within LinearNodalConstraint need to be shifted by -1 to make the result be correct. If using the node labels that are input, the result is wrong.

Below is a screenshot of both versions.
![image](https://github.com/idaholab/moose/assets/58032233/d2c81ecc-1fa1-4417-b95a-6e23f14179b6)

Also, attached are the files to reconstruct this problem.

Is this intended to shift nodes labels by -1, or is it a bug? 

The user might end up with very unintended results if the mesh in more complex involving many more nodes.

**Comment by GiudGiud** [comment]
Hello

The offset in the ids is expected. We have the same problem with paraview sometimes. 
`allow_renumbering=false` is for parallel / repartitioning renumbering, not the offset. 

If this is really impacting your workflow I would recommend either:
- adding a parameter to the FileMG to number as you expect
- creating a `RenumberNode/ElementIDMG` mesh generator object to perform this operation

Guillaume

**Reply by arndflatten** [reply]
Hi GiudGiud, thanks for your reply.

Do we understand correctly, that a Moose mesh internally always labels nodes from 0 to nNodes-1, but elements from 1 to nElems? Is this always the case, no matter which labels are defined in the input file that is fed to FileMeshGenerator? And will this convention remain or is this likely to be changed in some near future?

So, if read from an Abaqus *.inp mesh with consecutively numbered nodes and elements starting both from 1, all the nodes are relabelled internally by an offset of -1, but elementLabels remain unchanged - so any nodeLabel that is specified later in the inputfile (eg., to define constraints, outputs, etc) need to include this nodeNumber-shift?

Thanks a lot for clarifying, 
Arnd

**Reply by lindsayad** [reply, marked as answer]
in https://github.com/libMesh/libmesh/pull/3640 we made sure that we went from the 1-based numbering of Abaqus for both nodes and elements to 0-based, so you are half-correct. We translate the Abaqus numbering by -1 for both elements and nodes
